Lock down kwargs in offsets signatures

jbrockmendel opened the pull request:
Explicitly specify allowed kwargs in `__init__` for `Week`, `WeekOfMonth`, `LastWeekOfMonth` and `QuarterOffset`.

Ref #17176 

Intended to be orthogonal/complementary to #17450.

Will add tests after confirming we're on the same page.

- [ ] closes #xxxx
- [ ] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

jbrockmendel committed:
Lock down kwargs in offsets signatures

jreback reviewed (changes requested):
can you add a whatsnew note in apbreaking (maybe a sub-section), where you note which thinks are changing, and how it could potentitally affect the user. In future PR's can expand this as things are added.

jbrockmendel committed:
whatsnew note

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into offset_sigs

jreback commented on doc/source/whatsnew/v0.21.0.txt:
```diff
@@ -328,6 +328,13 @@ Previously, :func:`to_datetime` did not localize datetime ``Series`` data when `
+Restricting DateOffset Keywords
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Previously, ``DateOffset`` subclasses allowed arbitrary keyword arguments which could lead to unexpected behavior.  Now, subclasses :class:`Week`, :class:`WeekOfMonth`, :class:`LastWeekOfMonth`, :class:`QuarterOffset`, :class:`QuarterEnd`, :class:`QuarterBegin`, :class:`BQuarterEnd`, and :class:`BQuarterBegin`, will allow only specific arguments.  (:issue:`17176`).
```
aren't Month/MonthEnd included here? why not *all* classes?

jbrockmendel replied:
> aren't Month/MonthEnd included here?

Not as of now, no.

> why not _all_ classes?

Wanted to change them a couple at a time, err on the side of doing too little.  Now that #17486 is resolved I can push forward on a few more.  You want to just get them all in one go?

jreback replied:
yes for this type of change am ok with doing it in one fell swoop. On the whatsnew just highlite the change generally (e.g. give a single example), then list the classes that are affected (or if its all, then just say all). imagine a user who just 'uses' pandas. They want to quickly get to: does this change affect me, if so, what do I need to do to fix.

jreback reviewed (changes requested):
in-line comment

jreback commented:
ok, rebase this and i'll look again.

jbrockmendel commented:
Will rebase, but I haven't gotten around to addressing your earlier comments, so this one isn't ready yet.  #17564, #17585, #17363, #17543, and #17342 should be good to go.

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into offset_sigs
Lock down kwargs in Week.__init__

jbrockmendel committed:
restrict YearOffset to month kwarg

jbrockmendel committed:
Restrict SemiMonthOffset kwargs to day_of_month

jbrockmendel committed:
Lock down kwargs in FY5253 and FY5253Quarter

jbrockmendel committed:
lock down kwargs in BusinessDay, CustomBusinessMonthEnd

jbrockmendel committed:
Lockdown kwargs in remaining DateOffset subclasses

jbrockmendel committed:
Briefer WhatsNew, add missing kwds attrs

pep8speaks commented:
Hello @jbrockmendel! Thanks for updating the PR.

 - In the file [`pandas/tseries/offsets.py`](https://github.com/pandas-dev/pandas/blob/3daab6641cec3e418b4ea942bfb12f26e7cbc797/pandas/tseries/offsets.py), following are the PEP8 issues :

> [Line 2534:21](https://github.com/pandas-dev/pandas/blob/3daab6641cec3e418b4ea942bfb12f26e7cbc797/pandas/tseries/offsets.py#L2534): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 2535:21](https://github.com/pandas-dev/pandas/blob/3daab6641cec3e418b4ea942bfb12f26e7cbc797/pandas/tseries/offsets.py#L2535): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent

jbrockmendel commented:
OK, as requested, the new commits cover all of the existing classes.

jbrockmendel committed:
whitespace fixups

TomAugspurger reviewed (approved):
lgtm. Feel free to merge @jreback

jreback commented on doc/source/whatsnew/v0.21.0.txt:
```diff
@@ -387,6 +387,13 @@ New Behavior
 
+.. _whatsnew_0210.api.restricting_dateoffset_keywords
+
+Restricting DateOffset Keywords
```
ok, since we made this simpler, move to just a bullet point in Other API changes section.

jreback commented on pandas/tseries/offsets.py:
```diff
@@ -1121,7 +1125,7 @@ class SemiMonthOffset(DateOffset):
     _min_day_of_month = 2
 
-    def __init__(self, n=1, day_of_month=None, normalize=False, **kwds):
+    def __init__(self, n=1, normalize=False, day_of_month=None):
```
is this consistent with other freq signatures?

jbrockmendel replied:
All the other signatures are `(self, n=1, normalize=False, ...`.  In the status quo this is an outlier.

jreback replied:
yes that's fine. see my comment below though.

jreback replied:
this is fine, in an ideal world we would change it but can't now

jreback commented on pandas/tseries/offsets.py:
```diff
@@ -1742,21 +1747,22 @@ class LastWeekOfMonth(DateOffset):
     """
 
-    def __init__(self, n=1, normalize=False, **kwds):
+    def __init__(self, n=1, normalize=False, weekday=None):
```
is there a test for this as weekday=None? isn't that an error?

jbrockmendel replied:
It is an error, though a different one in Py2 vs Py3.  In the status quo not passing a weekday explicitly will raise a KeyError.  Default to Monday I guess?

jreback replied:
no if weekday is None, just raise a value error, in reality this should actually be the first (and a required parameter), but I guess that would be hard to change. Don't set defaults, raise

jbrockmendel replied:
Are you suggesting it should be `__init__(self, weekday, n=1, normalize=False)`?  I see some appeal to that, but there is a lot of symmetry with other signatures all starting with `(self, n=1, normalize=False, ...)`.

jreback replied:
see my comment above. this is ok for now. actually weekday is optional, so this is fine.

jbrockmendel replied:
Two things.  First, here and in other places you say "this is fine" but I don't know whether you're referring to the PR or the thing the PR is fixing.  Based on the go-ahead you've given below to look over timeseries.rst I'm assuming you mean the PR version is fine, just heads up for ways I'm liable to get confused next time around.

> actually weekday is optional,

Under the status quo, failing to pass `weekday` raises a `KeyError`.  In the PR, not passing anything will raise a ValueError in py2 and TypeError in py3.  (The same errors would result in the status quo if a user passed `None` explicitly)

jreback replied:
I never mean this entire PR is fine. If that was the case I wouldn't have any other comments. Individual comments are revering to the specific line/section where I make it.

jreback commented on pandas/tseries/offsets.py:
```diff
@@ -1829,13 +1835,14 @@ class QuarterOffset(DateOffset):
     #       point
 
-    def __init__(self, n=1, normalize=False, **kwds):
+    def __init__(self, n=1, normalize=False, startingMonth=None):
```
can you make an issue that this should be deprecated to ``starting_month`` (and prob others like this)

jbrockmendel committed:
Move doc section to other api changes bullet point

jbrockmendel committed:
Merge branch 'master' into offset_sigs

jreback reviewed (approved):
looks fine, can you pass thru the timeseries.rst docs and if anything needs changing there (should be ok), but who knows

jbrockmendel commented:
> looks fine, can you pass thru the timeseries.rst docs and if anything needs changing there (should be ok), but who knows

Didn't see anything that is affected by this PR, but a bunch of things that I want to double-check later given the outstanding GH issues.

jreback commented:
pls generate and add at least 1 file from ``generate_legacy_storage_files`` for pickle. I am not convinced this is backward compatible with existing pickles. It might be but can't determine without some tests.

jreback reviewed (changes requested):
see comments

jbrockmendel commented:
> pls generate and add at least 1 file from generate_legacy_storage_files for pickle. I am not convinced this is backward compatible with existing pickles. It might be but can't determine without some tests.

Following the instructions at the top of `tests.io.test_pickle`, installed 0.16.2, ran (0.20.3's)  "python generate_legacy_storage_files.py ~/Desktop/ pickle", got

```
Traceback (most recent call last):
  File "generate_legacy_storage_files.py", line 282, in <module>
    write_legacy_file()
  File "generate_legacy_storage_files.py", line 274, in write_legacy_file
    write_legacy_pickles(output_dir=output_dir)
  File "generate_legacy_storage_files.py", line 235, in write_legacy_pickles
    pickle.dump(create_pickle_data(), fh, pickle.HIGHEST_PROTOCOL)
  File "generate_legacy_storage_files.py", line 173, in create_pickle_data
    data = create_data()
  File "generate_legacy_storage_files.py", line 155, in create_data
    freq=Timestamp('2011-01-01', freq='D'),
  File "pandas/tslib.pyx", line 274, in pandas.tslib.Timestamp.__new__ (pandas/tslib.c:7571)
TypeError: __new__() got an unexpected keyword argument 'freq'
```

Will try again with 18ish.

Just to double-check: you're aware that test_offsets has `test_pickle_v0_15_2` and `test_pickle_compat_0_14_1`?

jbrockmendel commented:
Same with 0.18.1

jreback commented:
@jbrockmendel you may need to slightly modify the generation script to support back-generation. This is part of the challenge. you can use pandas versions to do that ONLY in that script to do that.  If you make changes then they should be checked in as part of this PR. 

generally we generate the most previous version or 2 with this script. Since you are touching offsets, need to go back and assure pickle compat for those.

jbrockmendel committed:
Add pickle tests for 0.19.2 and 0.20.3

jbrockmendel commented:
Added tests for 0.19.2 and 0.20.3, opened an issue to ask for clarification on these instructions in the docs.

jbrockmendel committed:
Merge branch 'offset_sigs' of https://github.com/jbrockmendel/pandas into offset_sigs

jbrockmendel commented:
circleCI log looks like the error is:

```
unable to execute 'x86_64-conda_cos6-linux-gnu-gcc': No such file or directory
error: command 'x86_64-conda_cos6-linux-gnu-gcc' failed with exit status 1
```

TomAugspurger commented:
If you rebase on master that error should go away. That was temporarily fixed in https://github.com/pandas-dev/pandas/pull/17700

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into offset_sigs

jreback commented on pandas/tests/tseries/data/dateoffset_0_19_2.pickle:
```diff
@@ -0,0 +1,187 @@
+S'YearBegin'
+p1
+ccopy_reg
+_reconstructor
```
not what i asked for at all

pls use the generate_legacy_data script
we have removed all of these ad hoc pickle things
we r not adding them back

jbrockmendel replied:
You asked me to run a script that doesn't work, then to troubleshoot it until it does.  I found existing tests and followed their example.  Please cut me some darn slack.

> we have removed all of these ad hoc pickle things

`test_pickle_v0_15_2` and `test_pickle_compat_0_14_1` might lead a naive reader (such as myself) to suspect otherwise.

> you can use pandas versions to do that ONLY in that script to do that. 

This sentence confuses me.

> If you make changes [to the generation script] then they should be checked in as part of this PR.

I have no interest in making changes to the non-functional generation script.

jreback replied:
> I have no interest in making changes to the non-functional generation script.

well it works fine according to the instructions I gave and has been used for quite a while.

> test_pickle_v0_15_2 and test_pickle_compat_0_14_1 might lead a naive reader

well if you look at what exists, the vast majority are in an organized format. The remaining ones should actually be converted to this format.
```
(pandas) bash-3.2$ find pandas -name '*pickle*'
pandas/compat/__pycache__/pickle_compat.cpython-36.pyc
pandas/compat/pickle_compat.py
pandas/io/__pycache__/pickle.cpython-36.pyc
pandas/io/pickle.py
pandas/tests/indexes/data/mindex_073.pickle
pandas/tests/indexes/data/multiindex_v1.pickle
pandas/tests/io/__pycache__/test_pickle.cpython-36-PYTEST.pyc
pandas/tests/io/data/categorical_0_14_1.pickle
pandas/tests/io/data/categorical_0_15_2.pickle
pandas/tests/io/data/legacy_pickle
pandas/tests/io/data/legacy_pickle/0.10.1/AMD64_windows_2.7.3.pickle
pandas/tests/io/data/legacy_pickle/0.10.1/x86_64_linux_2.7.3.pickle
pandas/tests/io/data/legacy_pickle/0.11.0/0.11.0_x86_64_linux_3.3.0.pickle
pandas/tests/io/data/legacy_pickle/0.11.0/x86_64_linux_2.7.3.pickle
pandas/tests/io/data/legacy_pickle/0.11.0/x86_64_linux_3.3.0.pickle
pandas/tests/io/data/legacy_pickle/0.12.0/0.12.0_AMD64_windows_2.7.3.pickle
pandas/tests/io/data/legacy_pickle/0.12.0/0.12.0_x86_64_linux_2.7.3.pickle
pandas/tests/io/data/legacy_pickle/0.13.0/0.13.0_AMD64_windows_2.7.3.pickle
pandas/tests/io/data/legacy_pickle/0.13.0/0.13.0_i686_linux_2.6.5.pickle
pandas/tests/io/data/legacy_pickle/0.13.0/0.13.0_i686_linux_2.7.3.pickle
pandas/tests/io/data/legacy_pickle/0.13.0/0.13.0_i686_linux_3.2.3.pickle
pandas/tests/io/data/legacy_pickle/0.13.0/0.13.0_x86_64_darwin_2.7.5.pickle
pandas/tests/io/data/legacy_pickle/0.13.0/0.13.0_x86_64_darwin_2.7.6.pickle
pandas/tests/io/data/legacy_pickle/0.13.0/0.13.0_x86_64_linux_2.7.3.pickle
pandas/tests/io/data/legacy_pickle/0.13.0/0.13.0_x86_64_linux_2.7.8.pickle
pandas/tests/io/data/legacy_pickle/0.13.0/0.13.0_x86_64_linux_3.3.0.pickle
pandas/tests/io/data/legacy_pickle/0.14.0/0.14.0_x86_64_darwin_2.7.6.pickle
pandas/tests/io/data/legacy_pickle/0.14.0/0.14.0_x86_64_linux_2.7.8.pickle
pandas/tests/io/data/legacy_pickle/0.14.1/0.14.1_x86_64_darwin_2.7.12.pickle
pandas/tests/io/data/legacy_pickle/0.14.1/0.14.1_x86_64_linux_2.7.8.pickle
pandas/tests/io/data/legacy_pickle/0.15.0/0.15.0_x86_64_darwin_2.7.12.pickle
pandas/tests/io/data/legacy_pickle/0.15.0/0.15.0_x86_64_linux_2.7.8.pickle
pandas/tests/io/data/legacy_pickle/0.15.2/0.15.2_x86_64_darwin_2.7.9.pickle
pandas/tests/io/data/legacy_pickle/0.16.0/0.16.0_x86_64_darwin_2.7.9.pickle
pandas/tests/io/data/legacy_pickle/0.16.2/0.16.2_AMD64_windows_2.7.10.pickle
pandas/tests/io/data/legacy_pickle/0.16.2/0.16.2_AMD64_windows_3.4.3.pickle
pandas/tests/io/data/legacy_pickle/0.16.2/0.16.2_x86_64_darwin_2.7.10.pickle
pandas/tests/io/data/legacy_pickle/0.16.2/0.16.2_x86_64_darwin_2.7.9.pickle
pandas/tests/io/data/legacy_pickle/0.16.2/0.16.2_x86_64_darwin_3.4.3.pickle
pandas/tests/io/data/legacy_pickle/0.16.2/0.16.2_x86_64_linux_2.7.10.pickle
pandas/tests/io/data/legacy_pickle/0.16.2/0.16.2_x86_64_linux_3.4.3.pickle
pandas/tests/io/data/legacy_pickle/0.17.0/0.17.0_AMD64_windows_2.7.11.pickle
pandas/tests/io/data/legacy_pickle/0.17.0/0.17.0_AMD64_windows_3.4.4.pickle
pandas/tests/io/data/legacy_pickle/0.17.0/0.17.0_x86_64_darwin_2.7.11.pickle
pandas/tests/io/data/legacy_pickle/0.17.0/0.17.0_x86_64_darwin_3.4.4.pickle
pandas/tests/io/data/legacy_pickle/0.17.0/0.17.0_x86_64_linux_2.7.11.pickle
pandas/tests/io/data/legacy_pickle/0.17.0/0.17.0_x86_64_linux_3.4.4.pickle
pandas/tests/io/data/legacy_pickle/0.17.0/0.17.1_AMD64_windows_2.7.11.pickle
pandas/tests/io/data/legacy_pickle/0.17.1/0.17.1_AMD64_windows_2.7.11.pickle
pandas/tests/io/data/legacy_pickle/0.17.1/0.17.1_x86_64_darwin_2.7.11.pickle
pandas/tests/io/data/legacy_pickle/0.18.0/0.18.0_AMD64_windows_2.7.11.pickle
pandas/tests/io/data/legacy_pickle/0.18.0/0.18.0_AMD64_windows_3.5.1.pickle
pandas/tests/io/data/legacy_pickle/0.18.0/0.18.0_x86_64_darwin_2.7.11.pickle
pandas/tests/io/data/legacy_pickle/0.18.0/0.18.0_x86_64_darwin_3.5.1.pickle
pandas/tests/io/data/legacy_pickle/0.18.1/0.18.1_x86_64_darwin_2.7.12.pickle
pandas/tests/io/data/legacy_pickle/0.18.1/0.18.1_x86_64_darwin_3.5.2.pickle
pandas/tests/io/data/legacy_pickle/0.19.2/0.19.2_x86_64_darwin_2.7.12.pickle
pandas/tests/io/data/legacy_pickle/0.19.2/0.19.2_x86_64_darwin_3.6.1.pickle
pandas/tests/io/test_pickle.py
pandas/tests/tseries/data/cday-0.14.1.pickle
pandas/tests/tseries/data/dateoffset_0_15_2.pickle
```

As you know pandas is a very large library. If things do not have process & procedure around them, then future generations will have pain. We have worked pretty hard (and you are contributing to this), to reduce technical debt.

jreback commented:
see https://github.com/pandas-dev/pandas/pull/17733

jreback commented:
@jbrockmendel 
since we really are locking down signatures. I would write an old (but acceptable signature) and add to ``generate_legacy_storage_files`` (e.g. maybe with the kwargs or something). I think this will all work, just want to confirm.

jreback commented:
I think I covered this case ``pandas/tests/tseries/data/dateoffset_0_15_2.pickle`` as well, so can remove that (now I didn't run it for 0.15.2 so you can certainly do that).

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into offset_sigs

jbrockmendel commented:
Thanks for fixing the script.  Added a file for 0.20.3 for kicks.

So since `test_pickle_v0_15_2` isn't an example for what you want a test to look like I'm guessing  iterate over the files unpickling them and checking that the `data['offsets']` matches the `off` dict defined in `generate_legacy_storage_files.create_data`?

> I think I covered this case `pandas/tests/tseries/data/dateoffset_0_15_2.pickle` as well, so can remove that (now I didn't run it for 0.15.2 so you can certainly do that).

There's some subject-verb ambiguity here.  What did you cover/run?  What can be removed?

> I would write an old (but acceptable signature) and add to generate_legacy_storage_files (e.g. maybe with the kwargs or something)

So go into `generate_legacy_storage_files` and add something to the `off` dict and generate a new pickle file?  But wouldn't that invalidate the existing files?

jbrockmendel committed:
New version of test_pickle_v0_20_3

jbrockmendel committed:
flake8 whitespace fixup

jreback commented:
> So go into generate_legacy_storage_files and add something to the off dict and generate a new pickle file? But wouldn't that invalidate the existing files?

no you can just generate a new one. We only compare to things that exist in the pickle AND in the generate script. So as long as you simply 'add' things its ok (e.g. don't rename or delete).

I added an offset section, you can simply add *moar* things to it (or even another version of it that say initializes with keyword args or somesuch).

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into offset_sigs

jbrockmendel commented:
OK, so I generated a new file by running the script under 0.20.3.  Then... what?  The existing pickle tests in tests.tseries.test_offsets are precedent you've explicitly said not to follow.

jreback commented on pandas/tests/tseries/test_offsets.py:
```diff
@@ -502,6 +502,31 @@ def test_pickle_v0_15_2(self):
         #
         tm.assert_dict_equal(offsets, read_pickle(pickle_path))
 
+    def test_pickle_v0_20_3(self):
```
none of this is needed

jreback replied:
these go in the generate_legacy_storage_files script instead. (which you then use to generate). I think the instructions are now pretty clear. If they are not, feel free to update them as well (at the top of that file).

jbrockmendel replied:
OK, just removed that test.  As to the instructions, I'm still unclear on where the actual _test_ goes and what it looks like.

jreback replied:
the test is generic it runs ALL of the pickle files

that’s the point you don’t actually need to construct a test just put a new pickle file

jbrockmendel replied:
OK.  `test_pickle_v0_20_3` has been removed, so it sounds like we're good to go.

jbrockmendel committed:
remove test per reviewer instruction

jreback reviewed (changes requested):
did you make changes in generate_legacy_storage_files? or just generate 0.20.3? I wanted you to update that with passing keywords so we can cover that as well, e.g.

ideally we would cover kwargs for most of the offsets.
```
In [3]: offsets.SemiMonthBegin(1, normalize=True, day_of_month=4)
Out[3]: <SemiMonthBegin: day_of_month=4>
```

This then assures that we have backcompat.

jbrockmendel committed:
Add offsets with kwds to generate_legacy_storage_files

jbrockmendel commented:
Just added a handful of new examples to `generate_legacy_storage_files`, updated the 0.20.3 file, created a 0.19.2 file.  Attempts to run under py3 raised

```
  File "pandas/pandas/tests/io/generate_legacy_storage_files.py", line 40, in <module>
    from pandas import (Series, DataFrame, Panel,
  File ".virtualenvs/pandas_0.19/lib/python3.6/site-packages/pandas/__init__.py", line 85, in <module>
    v = get_versions()
  File ".virtualenvs/pandas_0.19/lib/python3.6/site-packages/pandas/_version.py", line 23, in get_versions
    return json.loads(version_json)
AttributeError: module 'json' has no attribute 'loads'
```

jreback commented:
also seems we should update the doc-strings for the offsets now that the signature are ok, xref #8435

jreback commented:
>  File "pandas/pandas/tests/io/generate_legacy_storage_files.py", line 40, in <module>
    from pandas import (Series, DataFrame, Panel,
  File ".virtualenvs/pandas_0.19/lib/python3.6/site-packages/pandas/__init__.py", line 85, in <module>
    v = get_versions()
  File ".virtualenvs/pandas_0.19/lib/python3.6/site-packages/pandas/_version.py", line 23, in get_versions
    return json.loads(version_json)
AttributeError: module 'json' has no attribute 'loads'

This is because we have a top-level file called ``json.py`` for backwards compat (was a c-module),  which shadows the python json library, and the version things actually import this. So in this case you need to temporarily edit the ``pandas/_version.py`` file (in the version you are actually using) to avoid use of json (just change it to return a dict). a bit of a hack, but I have found no other way of fixing this (and it doesn't affect the released version, just running it like this).

jreback commented:
ok this looks fine.  Can you update the doc-strings? (also can do in another PR). now that the keywords are there they should be documented. lmk

jbrockmendel commented:
> Can you update the doc-strings? (also can do in another PR). 

This PR has been exhausting.  I'll put this on the TODO list.

jreback commented:
@jbrockmendel hahah, ok thanks1
